internals/servicelog/formatter.go

```diff
@@ -55,14 +51,79 @@ func NewFormatWriter(dest io.Writer, serviceName string) io.Writer {
 	}
 }
 
+// appendTimestamp appends a timestamp in format "YYYY-MM-DDTHH:mm:ss.sssZ" to
+// the given byte slice and returns the extended slice.
+//
+// The timestamp is always in UTC and has exactly 3 fractional digits
+// (millisecond precision). Makes no allocations if b has enough capacity.
+func appendTimestamp(b []byte, t time.Time) []byte {
+	const capacity = 24
+
+	utc := t.UTC()
+
+	year := utc.Year()
+	month := int(utc.Month())
+	day := utc.Day()
+	hour := utc.Hour()
+	minute := utc.Minute()
+	second := utc.Second()
+
+	// Convert nanoseconds to milliseconds as we use millisecond precision.
+	millisecond := utc.Nanosecond() / 1_000_000
+
+	// Ensure slice has enough capacity, and extend length.
+	b = slices.Grow(b, capacity)
+	b = b[:capacity]
+
+	// Write year (4 digits)
+	b[0] = byte('0' + year/1000%10)
+	b[1] = byte('0' + year/100%10)
+	b[2] = byte('0' + year/10%10)
+	b[3] = byte('0' + year%10)
+	b[4] = '-'
+
+	// Write month (2 digits)
+	b[5] = byte('0' + month/10)
+	b[6] = byte('0' + month%10)
+	b[7] = '-'
+
+	// Write day (2 digits)
+	b[8] = byte('0' + day/10)
+	b[9] = byte('0' + day%10)
+	b[10] = 'T'
+
+	// Write hour (2 digits)
+	b[11] = byte('0' + hour/10)
+	b[12] = byte('0' + hour%10)
+	b[13] = ':'
+
+	// Write minute (2 digits)
+	b[14] = byte('0' + minute/10)
+	b[15] = byte('0' + minute%10)
+	b[16] = ':'
+
+	// Write second (2 digits)
+	b[17] = byte('0' + second/10)
+	b[18] = byte('0' + second%10)
+	b[19] = '.'
+
+	// Write milliseconds (3 digits)
+	b[20] = byte('0' + millisecond/100%10)
```

Comment:
I think the range is fixed here, unlike years which can grow > 9999 ? So I would say the ```%10``` on this line can be dropped ?